plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java

```diff
@@ -0,0 +1,258 @@
+                    .isEqualTo(3L);
+
+            assertQueryFails(
+                    "ALTER TABLE " + table.getName() + " EXECUTE fast_forward('non-existing-branch', 'main')",
+                    "Branch 'non-existing-branch' does not exit");
+            assertQueryFails(
+                    "ALTER TABLE " + table.getName() + " EXECUTE fast_forward('main', 'non-existing-branch')",
+                    "Branch 'non-existing-branch' does not exit");
+        }
+    }
+
+    @Test
+    void testFastForwardNotAncestor()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_fast_forward", "(x int)")) {
+            createBranch(table.getName(), "test-branch");
+
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES 1", 1);
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') VALUES 1, 2, 3", 3);
+
+            assertQueryFails(
+                    "ALTER TABLE " + table.getName() + " EXECUTE fast_forward('main', 'test-branch')",
+                    "Branch 'main' is not an ancestor of 'test-branch'");
+        }
+    }
+
+    @Test
+    void testInsert()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_insert_into_branch", "(x int, y int)")) {
+            createBranch(table.getName(), "test-branch");
+
+            // insert into main (default) branch
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'main') VALUES (1, 2)", 1);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(1L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(0L);
+
+            // insert into another branch
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') VALUES (10, 20), (30, 40)", 2);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(1L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(2L);
+
+            // insert into another branch with a partial column
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') (x) VALUES 50", 1);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(1L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            // change table definition on main branch
+            assertUpdate("ALTER TABLE " + table.getName() + " ADD COLUMN z int");
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') SELECT x, y, NULL FROM " + table.getName(), 1);
```

Comment:
Are you sure that branch should be forced to be used with schema from current main? What if user drop column from main branch and we still want to add data for this column on branch? What if we want select with predicate on dropped column?